clang-tools-extra/clang-tidy/bugprone/IncorrectIteratorsCheck.cpp

```diff
@@ -0,0 +1,756 @@
+  auto MultiRangePotentiallyHalfOpen =
+      getMultiRangePotentiallyHalfOpenFunctors(getLangOpts());
+
+  auto MultiRangePolicySingleOutputIterator =
+      getMultiRangePolicyWithSingleOutputIterator(getLangOpts());
+
+  static const auto ToRefs =
+      [](std::initializer_list<std::reference_wrapper<std::vector<std::string>>>
+             Items) {
+        std::vector<StringRef> Result;
+        for (const auto &Item : Items) {
+          llvm::append_range(Result, Item.get());
+        }
+        return Result;
+      };
+
+  Finder->addMatcher(
+      makePairRangeMatcher(
+          ToRefs({SingleRange, SingleRangeHalf, SingleRangePolicy,
+                  SingleRangePolicyHalf, SingleRangeBackwardsHalf,
+                  MultiRangePolicy, MultiRangePotentiallyHalfOpen,
+                  MultiRangePolicyPotentiallyHalfOpen,
+                  MultiRangePolicySingleOutputIterator}),
+          0, 1, State),
+      this);
+  Finder->addMatcher(
+      makePairRangeMatcher(
+          ToRefs({SingleRangePolicy, SingleRangePolicyHalf, MultiRangePolicy,
+                  MultiRangePolicyPotentiallyHalfOpen,
+                  MultiRangePolicySingleOutputIterator}),
+          1, 2, State),
+      this);
+  Finder->addMatcher(
+      makeExpectedBeginFullMatcher(
+          ToRefs({SingleRangeHalf, SingleRangePolicyHalf}), 2, State),
+      this);
+  Finder->addMatcher(
+      makeExpectedBeginFullMatcher(ToRefs({SingleRangePolicyHalf}), 3, State),
+      this);
+  Finder->addMatcher(
+      makeExpectedEndFullMatcher(ToRefs({SingleRangeBackwardsHalf}), 2, State),
+      this);
+  Finder->addMatcher(
+      makePairRangeMatcher(ToRefs({MultiRangePolicy}), 2, 3, State), this);
+  Finder->addMatcher(
+      makePairRangeMatcher(ToRefs({MultiRangePolicy}), 3, 4, State), this);
+  Finder->addMatcher(
+      makeHalfOpenMatcher(ToRefs({MultiRangePotentiallyHalfOpen,
+                                  MultiRangePolicyPotentiallyHalfOpen,
+                                  MultiRangePolicySingleOutputIterator}),
+                          2, 3, State),
+      this);
+  Finder->addMatcher(
+      makeHalfOpenMatcher(ToRefs({MultiRangePolicyPotentiallyHalfOpen,
+                                  MultiRangePolicySingleOutputIterator}),
+                          3, 4, State),
+      this);
+  Finder->addMatcher(
+      makeExpectedBeginFullMatcher(
+          ToRefs({MultiRangePolicySingleOutputIterator}), 4, State),
+      this);
+  Finder->addMatcher(
+      makeExpectedBeginFullMatcher(
+          ToRefs({MultiRangePolicySingleOutputIterator}), 5, State),
+      this);
+
+  Finder->addMatcher(
+      makeContainerPairRangeMatcher({"::std::vector", "::std::list"},
+                                    {"insert"}, 1, 2, State),
+      this);
+  Finder->addMatcher(
+      makeContainerInternalMatcher({"::std::vector", "::std::list"},
+                                   {"insert", "erase", "emplace"}, 0, State),
+      this);
```

Comment:
There are some more container member that expect an iterator of the container, such as `std::deque::erase`. Can we add them now?

Reply:
Yeah, this is a little sore point, that there's not yet(#100349) a nice way to specify the available permutations